feat: remove `is-promise` library

EshaanAgg opened the pull request:
## Which problem is this PR solving?
Fixes #2074 

## Description of the changes
- Removed the `is-promise` library
- Added the type assertions wherever necessary
- Removed the redundant tests
- Made use of the native `Promise.allSettled` function (as described in a comment)

There seems to be an unrelated test in `date.test.js` that fails after these changes. I have fixed the test with what I think the intended testing was supposed to do. Would appreciate a review on the same!

## How was this change tested?
- Running the test suite locally

## Checklist
- [X] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [X] I have signed all commits
- [X] I have added unit tests for the new functionality
- [X] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

yurishkuro commented on packages/jaeger-ui/src/actions/jaeger-api.test.js:
```diff
@@ -49,11 +48,6 @@ describe('actions/jaeger-api', () => {
-  it('@JAEGER_API/FETCH_TRACE should return the promise', () => {
-    const { payload } = jaegerApiActions.fetchTrace(id);
-    expect(isPromise(payload)).toBeTruthy();
-  });
```
Why are these tests removed? What is replacing them?

yurishkuro replied:
> they are essentially type-checking, which should be TypeScript's job.

how is that type-checking achieved given that the source file in question is not in TS, but JS? 

Also

> Each of these fetchXYZ functions is essentially createAction wrapper 

That's an implementation detail of those functions, but the tests were enforcing the public API. Implementations may change.

EshaanAgg replied:
Ahh, my bad. I completely missed the `.js` extension. If we have to enforce the public API of these functions, then it would best to keep checking if we have a promise returned. I will revert the tests.

EshaanAgg commented on packages/jaeger-ui/src/actions/jaeger-api.test.js:
```diff
@@ -174,20 +133,3 @@ describe('actions/jaeger-api', () => {
   });
 });
-
-describe('allSettled', () => {
-  it('validate responses', async () => {
-    const res = await jaegerApiActions.allSettled([
-      Promise.resolve(1),
-      // eslint-disable-next-line prefer-promise-reject-errors
-      Promise.reject(2),
-      Promise.resolve(3),
-    ]);
-
-    expect(res).toEqual([
-      { status: 'fulfilled', value: 1 },
-      { status: 'rejected', reason: 2 },
-      { status: 'fulfilled', value: 3 },
-    ]);
-  });
```
I removed this test as we have now deprecated the custom implementation of `allSettled` for the builtin `Promise.allSettled`.

EshaanAgg commented:
I have reverted the tests testing the public API, and removed the `is-promise` library in favour of a simple custom implementation that checks the supplied object must be PromiseLike, that is, have a `then` and `catch` function. I have removed the analogous tests for the `readJsonFile` function as the file is in TS, and we can use explicit return type-assertions to make sure of the same.

yurishkuro commented:
Have you considered converting packages/jaeger-ui/src/actions/jaeger-api.js to Typescript?

EshaanAgg commented:
I tried to do so but later decided to avoid doing it for two reasons:
- We would have to go back to our custom implementation of `allSettled`
When changing the file to TS and using `allSettled`, we get the following TS error, asking us to increase the compiler to a more later one. Do you think this can lead to breaking changes?
```
Property 'allSettled' does not exist on type 'PromiseConstructor'. Do you need to change your target library? Try changing the 'lib' compiler option to 'es2020' or later. ts(2550)
```
- Loose types 
Even after flipping to TS, due to the nature of the `redux-actions` library, we get not very concrete types like the following:
```
const fetchAllServiceMetrics: ActionFunction<Action<any>, undefined, undefined, undefined, undefined>
const fetchDeepDependencyGraph: ActionFunctionAny<ActionMeta<any, {
    query: any;
}>>
const fetchServices: ActionFunction<Action<any>, undefined, undefined, undefined, undefined>
```
I don't think these types enforce the `PromiseLike` interface that you had earlier mentioned, thus making the exercise moot.

yurishkuro reviewed (approved):
Thanks!
